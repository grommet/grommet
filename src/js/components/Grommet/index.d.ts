import * as React from 'react';
import { ThemeType } from '../../themes';
import { BackgroundType } from '../../utils';

export interface GrommetProps {
  background?: BackgroundType;
  containerTarget?: HTMLElement;
  cssVars?: boolean;
  dir?: 'rtl';
  full?: boolean | 'min';
  options?: {
    box?: {
      cssGap?: boolean;
    };
    layer?: {
      singleId?: boolean;
    };
    drop?: {
      checkContainingBlock?: boolean;
    };
  };
  messages?: {
    messages?: {
      button?: {
        busy?: string;
        success?: string;
      };
      calendar?: {
        previousMove?: string;
        previous?: string;
        nextMove?: string;
        next?: string;
      };
      carousel?: {
        previous?: string;
        next?: string;
        jump?: string;
      };
      dateInput?: {
        openCalendar?: string;
        enterCalendar?: string;
        exitCalendar?: string;
      };
      dataChart?: {
        detailTitle?: string;
        detailFocus?: string;
      };
      dataFilters?: {
        clear?: string;
        heading?: string;
        open?: string;
        openSet?: {
          singular?: string;
          plural?: string;
        };
      };
      dataForm?: {
        submit?: string;
      };
      dataSearch?: {
        label?: string;
        open?: string;
      };
      dataSort?: {
        ascending?: string;
        by?: string;
        descending?: string;
        direction?: string;
        open?: string;
      };
      dataSummary?: {
        filtered?: string;
        filteredSingle?: string;
        items?: string;
        itemsSingle?: string;
        selected?: string;
        total?: string;
        totalSingle?: string;
      };
      dataTable?: {
        ascending?: string;
        collapse?: string;
        collapseAll?: string;
        decrease?: string;
        descending?: string;
        increase?: string;
        expand?: string;
        expandAll?: string;
        resizerAria?: string;
        rows?: string;
        rowsChanged?: string;
        rowsSingle?: string;
        searchBy?: string;
        total?: string;
        totalSingle?: string;
      };
      dataTableColumns?: {
        open?: string;
        order?: string;
        pinned?: string;
        select?: string;
        tip?: string;
        orderAria?: string;
        selectAria?: string;
      };
      dataTableGroupBy?: {
        clear?: string;
        label?: string;
      };
      dataView?: {
        label?: string;
      };
      fileInput?: {
        alert?: {
          maxFile?: string;
          maxSize?: string;
        };
        browse?: string;
        dropPrompt?: string;
        dropPromptMultiple?: string;
        files?: string;
        maxFile?: string;
        maxSizeSingle?: string;
        maxSizeMultiple?: {
          singular?: string;
          plural?: string;
        };
        remove?: string;
        removeAll?: string;
      };
      form?: {
        invalid?: string;
        required?: string;
      };
      formField?: {
        maxCharacters?: {
          remaining?: {
            singular?: string;
            plural?: string;
          };
          overLimit?: {
            singular?: string;
            plural: string;
          };
        };
      };
      list?: {
        pinned?: string;
      };
      menu?: {
        openMenu?: string;
        closeMenu?: string;
      };
      meter?: {
        bar?: {
          singular?: string;
          plural?: string;
        };
        circle?: {
          singular?: string;
          plural?: string;
        };
        pie?: {
          singular?: string;
          plural?: string;
        };
        semicircle?: {
          singular?: string;
          plural?: string;
        };
      };
      notification?: {
        close?: string;
      };
      pagination?: {
        stepLabel?: string;
        summary?: string;
        summaryNoItems?: string;
      };
      rangeSelector?: {
        lower?: string;
        upper?: string;
      };
      tag?: {
        removeLabel?: {
          nameAndValue?: string;
          valueOnly?: string;
        };
      };
      select?: {
        multiple?: string;
        selected?: string;
      };
      selectMultiple?: {
        clearAll?: string;
        clearAllA11y?: string;
        open?: string;
        optionNotSelected?: string;
        optionSelected?: string;
        search?: string;
        selectAll?: string;
        selectAllA11y?: string;
        selectDrop?: string;
        selected?: string;
        selectedOfTotal?: string;
        selectedOptions?: string;
        showMore?: string;
        summarizedValue?: string;
      };
      skipLinks?: {
        skipTo?: string;
      };
      tabs?: {
        tabContents?: string;
      };
      textInput?: {
        enterSelect?: string;
        suggestionsCount?: string;
        suggestionsExist?: string;
        suggestionIsOpen?: string;
      };
      video?: {
        audioDescriptions?: string;
        captions?: string;
        closeMenu?: string;
        description?: string;
        fullScreen?: string;
        progressMeter?: string;
        scrubber?: string;
        openMenu?: string;
        pauseButton?: string;
        playButton?: string;
        volumeDown?: string;
        volumeUp?: string;
      };
    };
    format?: (...args: any[]) => void;
  };
  onAnalytics?: (data: any) => void;
  plain?: boolean;
  theme?: ThemeType;
  themeMode?: 'dark' | 'light' | 'auto';
  userAgent?: string;
}

export interface GrommetExtendedProps
  extends GrommetProps,
    Omit<
      React.DetailedHTMLProps<
        React.HTMLAttributes<HTMLDivElement>,
        HTMLDivElement
      >,
      'dir'
    > {}

declare const Grommet: React.FC<GrommetExtendedProps>;

export { Grommet };
