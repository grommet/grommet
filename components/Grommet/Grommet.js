"use strict";

exports.__esModule = true;
exports.Grommet = void 0;
var _react = _interopRequireWildcard(require("react"));
var _styledComponents = require("styled-components");
var _contexts = require("../../contexts");
var _utils = require("../../utils");
var _themes = require("../../themes");
var _StyledGrommet = require("./StyledGrommet");
var _RootsContext = require("../../contexts/RootsContext");
var _OptionsContext = require("../../contexts/OptionsContext");
var _MessageContext = require("../../contexts/MessageContext");
var _default = _interopRequireDefault(require("../../languages/default.json"));
var _propTypes = require("./propTypes");
var _AnalyticsContext = require("../../contexts/AnalyticsContext");
var _excluded = ["children", "full", "containerTarget", "theme", "options", "messages", "onAnalytics"];
function _interopRequireDefault(e) { return e && e.__esModule ? e : { "default": e }; }
function _getRequireWildcardCache(e) { if ("function" != typeof WeakMap) return null; var r = new WeakMap(), t = new WeakMap(); return (_getRequireWildcardCache = function _getRequireWildcardCache(e) { return e ? t : r; })(e); }
function _interopRequireWildcard(e, r) { if (!r && e && e.__esModule) return e; if (null === e || "object" != typeof e && "function" != typeof e) return { "default": e }; var t = _getRequireWildcardCache(r); if (t && t.has(e)) return t.get(e); var n = { __proto__: null }, a = Object.defineProperty && Object.getOwnPropertyDescriptor; for (var u in e) if ("default" !== u && {}.hasOwnProperty.call(e, u)) { var i = a ? Object.getOwnPropertyDescriptor(e, u) : null; i && (i.get || i.set) ? Object.defineProperty(n, u, i) : n[u] = e[u]; } return n["default"] = e, t && t.set(e, n), n; }
function _extends() { return _extends = Object.assign ? Object.assign.bind() : function (n) { for (var e = 1; e < arguments.length; e++) { var t = arguments[e]; for (var r in t) ({}).hasOwnProperty.call(t, r) && (n[r] = t[r]); } return n; }, _extends.apply(null, arguments); }
function _objectWithoutPropertiesLoose(r, e) { if (null == r) return {}; var t = {}; for (var n in r) if ({}.hasOwnProperty.call(r, n)) { if (-1 !== e.indexOf(n)) continue; t[n] = r[n]; } return t; }
var FullGlobalStyle = (0, _styledComponents.createGlobalStyle)(["body{margin:0;}"]);
var defaultOptions = {};
var Grommet = exports.Grommet = /*#__PURE__*/(0, _react.forwardRef)(function (props, ref) {
  var children = props.children,
    full = props.full,
    _props$containerTarge = props.containerTarget,
    containerTarget = _props$containerTarge === void 0 ? typeof document === 'object' ? document.body : undefined : _props$containerTarge,
    themeProp = props.theme,
    _props$options = props.options,
    options = _props$options === void 0 ? defaultOptions : _props$options,
    messagesProp = props.messages,
    onAnalytics = props.onAnalytics,
    rest = _objectWithoutPropertiesLoose(props, _excluded);
  var background = props.background,
    dir = props.dir,
    themeMode = props.themeMode,
    userAgent = props.userAgent;
  var _useState = (0, _react.useState)(),
    stateResponsive = _useState[0],
    setResponsive = _useState[1];
  var theme = (0, _react.useMemo)(function () {
    var nextTheme = (0, _utils.deepMerge)(_themes.base, themeProp || {});

    // if user provides specific menu alignment, we don't want
    // the defaults to be included at all (can cause issues with controlMirror)
    // override merged value with themeProp value
    if (themeProp && themeProp.menu && themeProp.menu.drop && themeProp.menu.drop.align) {
      delete nextTheme.menu.drop.align;
      nextTheme.menu.drop.align = themeProp.menu.drop.align;
    }
    var themeBackground = nextTheme.global.colors.background;
    nextTheme.dark = (themeMode || nextTheme.defaultMode) === 'dark';
    if (themeMode === 'auto' && typeof window !== 'undefined' && window.matchMedia && window.matchMedia('(prefers-color-scheme: dark)').matches) {
      nextTheme.dark = true;
    }
    var color = (0, _utils.normalizeColor)(background || themeBackground, nextTheme);
    nextTheme.dark = (0, _utils.backgroundIsDark)(color, nextTheme);
    nextTheme.baseBackground = background || themeBackground;
    // This allows DataTable to intelligently set the background of a pinned
    // header or footer.
    nextTheme.background = nextTheme.baseBackground;
    if (dir) {
      nextTheme.dir = dir;
    }
    return nextTheme;
  }, [background, dir, themeMode, themeProp]);
  var messages = (0, _react.useMemo)(function () {
    // combine the passed in messages, if any, with the default
    // messages and format function.
    var nextMessages = (0, _utils.deepMerge)(_default["default"], (messagesProp == null ? void 0 : messagesProp.messages) || {});
    return {
      messages: nextMessages,
      format: function format(opts) {
        var message = (messagesProp == null ? void 0 : messagesProp.format) && messagesProp.format(opts);
        return typeof message !== 'undefined' ? message : (0, _MessageContext.format)(opts, nextMessages);
      }
    };
  }, [messagesProp]);
  (0, _react.useEffect)(function () {
    var onResize = function onResize() {
      setResponsive((0, _utils.getBreakpoint)(document.body.clientWidth, theme));
    };
    window.addEventListener('resize', onResize);
    onResize();
    return function () {
      window.removeEventListener('resize', onResize);
    };
  }, [theme]);
  var responsive = stateResponsive || (0, _utils.deviceResponsive)(userAgent, theme) || theme.global.deviceBreakpoints.tablet;
  var grommetRef = (0, _utils.useForwardedRef)(ref);

  // track open FocusedContainers in a global array to manage
  // focus event listeners for trapFocus
  var roots = (0, _react.useRef)([]);
  (0, _react.useEffect)(function () {
    if (grommetRef.current) roots.current.push(grommetRef.current);
  }, [grommetRef]);
  var rootsContextValue = (0, _react.useMemo)(function () {
    return {
      roots: roots
    };
  }, []);
  return /*#__PURE__*/_react["default"].createElement(_contexts.ThemeContext.Provider, {
    value: theme
  }, /*#__PURE__*/_react["default"].createElement(_contexts.ResponsiveContext.Provider, {
    value: responsive
  }, /*#__PURE__*/_react["default"].createElement(_RootsContext.RootsContext.Provider, {
    value: rootsContextValue
  }, /*#__PURE__*/_react["default"].createElement(_contexts.ContainerTargetContext.Provider, {
    value: containerTarget
  }, /*#__PURE__*/_react["default"].createElement(_OptionsContext.OptionsContext.Provider, {
    value: options
  }, /*#__PURE__*/_react["default"].createElement(_MessageContext.MessageContext.Provider, {
    value: messages
  }, /*#__PURE__*/_react["default"].createElement(_AnalyticsContext.AnalyticsProvider, {
    onAnalytics: onAnalytics
  }, /*#__PURE__*/_react["default"].createElement(_StyledGrommet.StyledGrommet, _extends({
    full: full
  }, rest, {
    ref: grommetRef
  }), children), full && /*#__PURE__*/_react["default"].createElement(FullGlobalStyle, null))))))));
});
Grommet.displayName = 'Grommet';
Grommet.propTypes = _propTypes.GrommetPropTypes;