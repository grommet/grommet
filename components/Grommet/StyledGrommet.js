"use strict";

exports.__esModule = true;
exports.StyledGrommet = void 0;
var _styledComponents = _interopRequireWildcard(require("styled-components"));
var _utils = require("../../utils");
function _interopRequireWildcard(e, t) { if ("function" == typeof WeakMap) var r = new WeakMap(), n = new WeakMap(); return (_interopRequireWildcard = function _interopRequireWildcard(e, t) { if (!t && e && e.__esModule) return e; var o, i, f = { __proto__: null, "default": e }; if (null === e || "object" != typeof e && "function" != typeof e) return f; if (o = t ? n : r) { if (o.has(e)) return o.get(e); o.set(e, f); } for (var _t in e) "default" !== _t && {}.hasOwnProperty.call(e, _t) && ((i = (o = Object.defineProperty) && Object.getOwnPropertyDescriptor(e, _t)) && (i.get || i.set) ? o(f, _t, i) : f[_t] = e[_t]); return f; })(e, t); }
var fullStyle = function fullStyle(full) {
  if (full === 'min') return (0, _styledComponents.css)(["min-height:100vh;"]);
  return (0, _styledComponents.css)(["width:100vw;height:100vh;overflow:auto;"]);
};
var StyledGrommet = exports.StyledGrommet = _styledComponents["default"].div.withConfig(_utils.styledComponentsConfig).withConfig({
  displayName: "StyledGrommet",
  componentId: "sc-19lkkz7-0"
})(["", " ", " ", " ", " ", ""], function (props) {
  return !props.plain && _utils.baseStyle;
}, function (props) {
  return props.full && fullStyle(props.full);
}, function (props) {
  return props.theme.global.font.face;
}, function (props) {
  return props.theme.grommet.extend;
}, function (props) {
  return props.cssVars && Object.keys(props.theme.global.colors).filter(function (k) {
    return typeof props.theme.global.colors[k] === 'string';
  }).map(function (k) {
    return "--" + k + ": " + props.theme.global.colors[k] + ";";
  }).join('\n');
});