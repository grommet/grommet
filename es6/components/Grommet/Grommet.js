var _excluded = ["children", "full", "containerTarget", "theme", "options", "messages", "onAnalytics"];
function _extends() { return _extends = Object.assign ? Object.assign.bind() : function (n) { for (var e = 1; e < arguments.length; e++) { var t = arguments[e]; for (var r in t) ({}).hasOwnProperty.call(t, r) && (n[r] = t[r]); } return n; }, _extends.apply(null, arguments); }
function _objectWithoutPropertiesLoose(r, e) { if (null == r) return {}; var t = {}; for (var n in r) if ({}.hasOwnProperty.call(r, n)) { if (-1 !== e.indexOf(n)) continue; t[n] = r[n]; } return t; }
import React, { forwardRef, useEffect, useMemo, useRef, useState } from 'react';
import { createGlobalStyle } from 'styled-components';
import { ContainerTargetContext, ResponsiveContext, ThemeContext } from '../../contexts';
import { deepMerge, backgroundIsDark, deviceResponsive, getBreakpoint, normalizeColor, useForwardedRef } from '../../utils';
import { base as baseTheme } from '../../themes';
import { StyledGrommet } from './StyledGrommet';
import { RootsContext } from '../../contexts/RootsContext';
import { OptionsContext } from '../../contexts/OptionsContext';
import { format as _format, MessageContext } from '../../contexts/MessageContext';
import defaultMessages from '../../languages/default.json';
import { GrommetPropTypes } from './propTypes';
import { AnalyticsProvider } from '../../contexts/AnalyticsContext';
var FullGlobalStyle = createGlobalStyle(["body{margin:0;}"]);
var defaultOptions = {};
var Grommet = /*#__PURE__*/forwardRef(function (props, ref) {
  var children = props.children,
    full = props.full,
    _props$containerTarge = props.containerTarget,
    containerTarget = _props$containerTarge === void 0 ? typeof document === 'object' ? document.body : undefined : _props$containerTarge,
    themeProp = props.theme,
    _props$options = props.options,
    options = _props$options === void 0 ? defaultOptions : _props$options,
    messagesProp = props.messages,
    onAnalytics = props.onAnalytics,
    rest = _objectWithoutPropertiesLoose(props, _excluded);
  var background = props.background,
    dir = props.dir,
    themeMode = props.themeMode,
    userAgent = props.userAgent;
  var _useState = useState(),
    stateResponsive = _useState[0],
    setResponsive = _useState[1];
  var theme = useMemo(function () {
    var nextTheme = deepMerge(baseTheme, themeProp || {});

    // if user provides specific menu alignment, we don't want
    // the defaults to be included at all (can cause issues with controlMirror)
    // override merged value with themeProp value
    if (themeProp && themeProp.menu && themeProp.menu.drop && themeProp.menu.drop.align) {
      delete nextTheme.menu.drop.align;
      nextTheme.menu.drop.align = themeProp.menu.drop.align;
    }
    var themeBackground = nextTheme.global.colors.background;
    nextTheme.dark = (themeMode || nextTheme.defaultMode) === 'dark';
    if (themeMode === 'auto' && typeof window !== 'undefined' && window.matchMedia && window.matchMedia('(prefers-color-scheme: dark)').matches) {
      nextTheme.dark = true;
    }
    var color = normalizeColor(background || themeBackground, nextTheme);
    nextTheme.dark = backgroundIsDark(color, nextTheme);
    nextTheme.baseBackground = background || themeBackground;
    // This allows DataTable to intelligently set the background of a pinned
    // header or footer.
    nextTheme.background = nextTheme.baseBackground;
    if (dir) {
      nextTheme.dir = dir;
    }
    return nextTheme;
  }, [background, dir, themeMode, themeProp]);
  var messages = useMemo(function () {
    // combine the passed in messages, if any, with the default
    // messages and format function.
    var nextMessages = deepMerge(defaultMessages, (messagesProp == null ? void 0 : messagesProp.messages) || {});
    return {
      messages: nextMessages,
      format: function format(opts) {
        var message = (messagesProp == null ? void 0 : messagesProp.format) && messagesProp.format(opts);
        return typeof message !== 'undefined' ? message : _format(opts, nextMessages);
      }
    };
  }, [messagesProp]);
  useEffect(function () {
    var onResize = function onResize() {
      setResponsive(getBreakpoint(document.body.clientWidth, theme));
    };
    window.addEventListener('resize', onResize);
    onResize();
    return function () {
      window.removeEventListener('resize', onResize);
    };
  }, [theme]);
  var responsive = stateResponsive || deviceResponsive(userAgent, theme) || theme.global.deviceBreakpoints.tablet;
  var grommetRef = useForwardedRef(ref);

  // track open FocusedContainers in a global array to manage
  // focus event listeners for trapFocus
  var roots = useRef([]);
  useEffect(function () {
    if (grommetRef.current) roots.current.push(grommetRef.current);
  }, [grommetRef]);
  var rootsContextValue = useMemo(function () {
    return {
      roots: roots
    };
  }, []);
  return /*#__PURE__*/React.createElement(ThemeContext.Provider, {
    value: theme
  }, /*#__PURE__*/React.createElement(ResponsiveContext.Provider, {
    value: responsive
  }, /*#__PURE__*/React.createElement(RootsContext.Provider, {
    value: rootsContextValue
  }, /*#__PURE__*/React.createElement(ContainerTargetContext.Provider, {
    value: containerTarget
  }, /*#__PURE__*/React.createElement(OptionsContext.Provider, {
    value: options
  }, /*#__PURE__*/React.createElement(MessageContext.Provider, {
    value: messages
  }, /*#__PURE__*/React.createElement(AnalyticsProvider, {
    onAnalytics: onAnalytics
  }, /*#__PURE__*/React.createElement(StyledGrommet, _extends({
    full: full
  }, rest, {
    ref: grommetRef
  }), children), full && /*#__PURE__*/React.createElement(FullGlobalStyle, null))))))));
});
Grommet.displayName = 'Grommet';
Grommet.propTypes = GrommetPropTypes;
export { Grommet };