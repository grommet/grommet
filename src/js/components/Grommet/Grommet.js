import React, { forwardRef, useEffect, useMemo, useRef, useState } from 'react';
import { createGlobalStyle } from 'styled-components';

import {
  ContainerTargetContext,
  ResponsiveContext,
  ThemeContext,
} from '../../contexts';

import {
  deepMerge,
  backgroundIsDark,
  deviceResponsive,
  getBreakpoint,
  normalizeColor,
  useForwardedRef,
} from '../../utils';
import { base as baseTheme } from '../../themes';
import { StyledGrommet } from './StyledGrommet';
import { RootsContext } from '../../contexts/RootsContext';
import { OptionsContext } from '../../contexts/OptionsContext';
import { format, MessageContext } from '../../contexts/MessageContext';
import defaultMessages from '../../languages/default.json';
import { GrommetPropTypes } from './propTypes';
import { AnalyticsProvider } from '../../contexts/AnalyticsContext';

const FullGlobalStyle = createGlobalStyle`
  body { margin: 0; }
`;

const defaultOptions = {};

const Grommet = forwardRef((props, ref) => {
  const {
    children,
    full,
    containerTarget = typeof document === 'object' ? document.body : undefined,
    theme: themeProp,
    options = defaultOptions,
    messages: messagesProp,
    onAnalytics,
    ...rest
  } = props;
  const { background, dir, themeMode, userAgent } = props;
  const [stateResponsive, setResponsive] = useState();

  const theme = useMemo(() => {
    const nextTheme = deepMerge(baseTheme, themeProp || {});

    // if user provides specific menu alignment, we don't want
    // the defaults to be included at all (can cause issues with controlMirror)
    // override merged value with themeProp value
    if (
      themeProp &&
      themeProp.menu &&
      themeProp.menu.drop &&
      themeProp.menu.drop.align
    ) {
      delete nextTheme.menu.drop.align;
      nextTheme.menu.drop.align = themeProp.menu.drop.align;
    }
    const {
      colors: { background: themeBackground },
    } = nextTheme.global;

    nextTheme.dark = (themeMode || nextTheme.defaultMode) === 'dark';

    if (
      themeMode === 'auto' &&
      typeof window !== 'undefined' &&
      window.matchMedia &&
      window.matchMedia('(prefers-color-scheme: dark)').matches
    ) {
      nextTheme.dark = true;
    }

    const color = normalizeColor(background || themeBackground, nextTheme);
    nextTheme.dark = backgroundIsDark(color, nextTheme);
    nextTheme.baseBackground = background || themeBackground;
    // This allows DataTable to intelligently set the background of a pinned
    // header or footer.
    nextTheme.background = nextTheme.baseBackground;

    if (dir) {
      nextTheme.dir = dir;
    }

    return nextTheme;
  }, [background, dir, themeMode, themeProp]);

  const messages = useMemo(() => {
    // combine the passed in messages, if any, with the default
    // messages and format function.
    const nextMessages = deepMerge(
      defaultMessages,
      messagesProp?.messages || {},
    );
    return {
      messages: nextMessages,
      format: (opts) => {
        const message = messagesProp?.format && messagesProp.format(opts);
        return typeof message !== 'undefined'
          ? message
          : format(opts, nextMessages);
      },
    };
  }, [messagesProp]);

  useEffect(() => {
    const onResize = () => {
      setResponsive(getBreakpoint(document.body.clientWidth, theme));
    };
    window.addEventListener('resize', onResize);
    onResize();
    return () => {
      window.removeEventListener('resize', onResize);
    };
  }, [theme]);

  const responsive =
    stateResponsive ||
    deviceResponsive(userAgent, theme) ||
    theme.global.deviceBreakpoints.tablet;

  const grommetRef = useForwardedRef(ref);

  // track open FocusedContainers in a global array to manage
  // focus event listeners for trapFocus
  const roots = useRef([]);
  useEffect(() => {
    if (grommetRef.current) roots.current.push(grommetRef.current);
  }, [grommetRef]);
  const rootsContextValue = useMemo(() => ({ roots }), []);

  return (
    <ThemeContext.Provider value={theme}>
      <ResponsiveContext.Provider value={responsive}>
        <RootsContext.Provider value={rootsContextValue}>
          <ContainerTargetContext.Provider value={containerTarget}>
            <OptionsContext.Provider value={options}>
              <MessageContext.Provider value={messages}>
                <AnalyticsProvider onAnalytics={onAnalytics}>
                  <StyledGrommet full={full} {...rest} ref={grommetRef}>
                    {children}
                  </StyledGrommet>
                  {full && <FullGlobalStyle />}
                </AnalyticsProvider>
              </MessageContext.Provider>
            </OptionsContext.Provider>
          </ContainerTargetContext.Provider>
        </RootsContext.Provider>
      </ResponsiveContext.Provider>
    </ThemeContext.Provider>
  );
});

Grommet.displayName = 'Grommet';
Grommet.propTypes = GrommetPropTypes;

export { Grommet };
