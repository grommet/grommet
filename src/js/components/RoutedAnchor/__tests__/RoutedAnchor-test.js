import React, { useMemo } from 'react';
import PropTypes from 'prop-types';
import { render, screen, fireEvent, createEvent } from '@testing-library/react';
import userEvent from '@testing-library/user-event';
import 'jest-styled-components';
import 'regenerator-runtime/runtime';

import { Grommet } from '../../Grommet';
import { RoutedAnchor } from '..';
import { RouterContext } from '../RouterContext';

const FakeRouter = ({ children, push, replace }) => {
  const value = useMemo(
    () => ({
      history: {
        push,
        replace,
      },
    }),
    [push, replace],
  );

  return (
    <RouterContext.Provider value={value}>
      <div>{children}</div>
    </RouterContext.Provider>
  );
};

FakeRouter.propTypes = {
  children: PropTypes.node.isRequired,
  push: PropTypes.func.isRequired,
  replace: PropTypes.func.isRequired,
};

describe('RoutedAnchor', () => {
  const replace = jest.fn();
  const push = jest.fn();
  let warnSpy;

  beforeEach(() => {
    warnSpy = jest.spyOn(console, 'warn').mockImplementation();
  });

  afterEach(() => {
    warnSpy.mockRestore();
  });

  test('renders', () => {
    const { container } = render(
      <Grommet>
        <FakeRouter push={push} replace={replace}>
          <RoutedAnchor label="Test" path="/" />
        </FakeRouter>
      </Grommet>,
    );

    expect(warnSpy).toHaveBeenCalled();
    expect(container.firstChild).toMatchSnapshot();
  });

  test('is clickable', () => {
    const preventDefault = jest.fn();
    const onClick = jest.fn();
    render(
      <Grommet>
        <FakeRouter push={push} replace={replace}>
          <RoutedAnchor label="Test" onClick={onClick} path="/" />
        </FakeRouter>
      </Grommet>,
    );

    const anchor = screen.getByRole('link');

    const clickEvent = createEvent.click(anchor);
    clickEvent.preventDefault = preventDefault;
    fireEvent(anchor, clickEvent);

    expect(onClick).toHaveBeenCalledTimes(1);
    expect(push).toHaveBeenCalledTimes(1);
    expect(preventDefault).toHaveBeenCalledTimes(1);
    expect(warnSpy).toHaveBeenCalled();
  });

  test('skips onClick if right clicked', async () => {
    const user = userEvent.setup();

    const onClick = jest.fn();
    render(
      <Grommet>
        <FakeRouter push={push} replace={replace}>
          <RoutedAnchor label="Test" onClick={onClick} path="/" />
        </FakeRouter>
      </Grommet>,
    );

    const anchor = screen.getByRole('link');

    await user.pointer([
      { target: anchor },
      { keys: '[MouseRight]', target: anchor },
    ]);

    expect(onClick).not.toHaveBeenCalled();
    expect(warnSpy).toHaveBeenCalled();
  });

  test('calls router context push', () => {
    const preventDefault = jest.fn();
    render(
      <Grommet>
        <FakeRouter push={push} replace={replace}>
          <RoutedAnchor label="Test" path="/" />
        </FakeRouter>
      </Grommet>,
    );

    const anchor = screen.getByRole('link');

    const clickEvent = createEvent.click(anchor);
    clickEvent.preventDefault = preventDefault;
    fireEvent(anchor, clickEvent);

    expect(preventDefault).toHaveBeenCalled();
    expect(push).toHaveBeenCalledWith('/');
    expect(warnSpy).toHaveBeenCalled();
  });

  test('calls router context replace', () => {
    const preventDefault = jest.fn();
    render(
      <Grommet>
        <FakeRouter replace={replace} push={push}>
          <RoutedAnchor label="Test" path="/" method="replace" />
        </FakeRouter>
      </Grommet>,
    );

    const anchor = screen.getByRole('link');

    const clickEvent = createEvent.click(anchor);
    clickEvent.preventDefault = preventDefault;
    fireEvent(anchor, clickEvent);

    expect(preventDefault).toHaveBeenCalled();
    expect(replace).toHaveBeenCalledWith('/');
    expect(warnSpy).toHaveBeenCalled();
  });
});
