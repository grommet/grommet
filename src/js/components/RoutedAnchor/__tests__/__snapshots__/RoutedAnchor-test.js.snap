// Jest Snapshot v1, https://jestjs.io/docs/snapshot-testing

exports[`RoutedAnchor renders 1`] = `
.c0 {
  font-size: 18px;
  line-height: 24px;
  box-sizing: border-box;
  -webkit-text-size-adjust: 100%;
  -ms-text-size-adjust: 100%;
  -moz-osx-font-smoothing: grayscale;
  -webkit-font-smoothing: antialiased;
}

.c1 {
  box-sizing: border-box;
  display: inline-flex;
  font-size: inherit;
  line-height: inherit;
  color: #7D4CDB;
  font-weight: 600;
  text-decoration: none;
  cursor: pointer;
}

.c1:hover {
  text-decoration: underline;
}

<div
  class="c0"
>
  <div>
    <a
      class="c1"
      href="/"
    >
      Test
    </a>
  </div>
</div>
`;
